prov/sockets/src/sock_progress.c

```diff
@@ -2629,10 +2630,12 @@ static void sock_pe_poll(struct sock_pe *pe)
 	if (pe->waittime && ((fi_gettime_ms() - pe->waittime) < sock_pe_waittime))
 		return;
 
-	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list))
+	pthread_mutex_lock(&pe->list_lock);
+	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list)) {
+		pthread_mutex_unlock(&pe->list_lock);
 		return;
+	}
 
-	pthread_mutex_lock(&pe->list_lock);
 	if (!dlist_empty(&pe->tx_list)) {
```

Comment:
The issue I'm trying to avoid is acquiring and releasing the same mutex multiple times.  Even though we're running sockets, this is in performance related code -- the progress thread.  The original code did not acquire the mutex if there was work queued.  The new code does, then releases it, only to immediately re-acquire it.  I'm suggesting that we make helgrind happy without introducing more overhead.